cluster/addons/dns/kube2sky/tree_test.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	pathlib "path"
+	"strings"
+	"testing"
+)
+
+func makeRecordTree(kvs []kvPair) (*recordTree, error) {
+	var records recordTree
+	for _, kv := range kvs {
+		if err := records.Set(kv.k, kv.v); err != nil {
+			return nil, fmt.Errorf("error while setting %s to %s: %v", kv.k, kv.v, err)
+		}
+	}
+	return &records, nil
+}
+
+type kvPair struct {
+	k string
+	v string
+}
+
+func newKVPairs(kv ...string) []kvPair {
+	var ps []kvPair
+	for i := 0; i < len(kv); i += 2 {
+		ps = append(ps, kvPair{kv[i], kv[i+1]})
+	}
+	return ps
+}
+
+func TestRecordTree(t *testing.T) {
+	cases := []struct {
+		kv  []kvPair
+		err bool
+	}{
+		{newKVPairs("/a", "`", "/b", "1", "/c", "2"), false},
+		{newKVPairs("/a/b", "2", "/a/c", "][p"), false},
+		{newKVPairs("/a/b/c", "3512123491234", "/a/r", "2"), false},
+		{newKVPairs("/skydns/a", "x", "/skydns/b/c/d", "y", "/skydns/b/c/e", "z", "/skydns/b/c/f", "2"), false},
+		{newKVPairs("/a/b", "2", "/a/b", "3"), false},
+		{newKVPairs("/a/", "2"), true},
+		{newKVPairs("/skydns/a", "3", "/skydns/b", "$", "/skydns/a/f", "1"), true},
+		{newKVPairs("/skydns/a/c", "3", "/skydns/b", "$", "/skydns/a", "1"), true},
+	}
+	for i, test := range cases {
+		_, err := makeRecordTree(test.kv)
+		if test.err {
+			if err == nil {
+				t.Errorf("case %d: expected error", i)
+			}
+		} else if err != nil {
+			t.Errorf("case %d: unexpected error %v", i, err)
+		}
+	}
+}
+
+func addMissingKV(kvs []kvPair) []kvPair {
+	newMap := make(map[string]string)
+	for _, kv := range kvs {
+		split := strings.Split(kv.k, "/")
+		for i := 1; i < len(split); i++ {
+			newMap["/"+pathlib.Join(split[:i]...)] = ""
+		}
+		newMap[kv.k] = kv.v
+	}
+	var ps []kvPair
+	for k, v := range newMap {
+		ps = append(ps, kvPair{k, v})
+	}
+	return ps
+}
+
+func (rs flatRecords) sameAsKV(kvs []kvPair) error {
+	if len(rs) != len(kvs) {
+		return fmt.Errorf("different lengths")
+	}
+	for _, r := range rs {
+		var found bool
+		for _, kv := range kvs {
+			if r.key == kv.k {
+				if r.IsDir() {
```

Comment:
Removed the method call, but otherwise this looks like it's doing fairly dumb comparisons. I'm not sure how much value there is in this test. It's comparing populating a recordTree and flattening it to taking a different flat representation and then populating it with directory entries. It's basically doing the same thing twice but using a different (and slightly simpler representation) to validate the real implementation.